File: src/main/java/com/github/njuro/jard/user/UserResolver.java
package com.github.njuro.jard.user;

import com.github.njuro.jard.common.Mappings;
import com.github.njuro.jard.user.dto.UserDto;
import com.github.njuro.jard.utils.PathVariableArgumentResolver;
import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.springframework.core.MethodParameter;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.ModelAndViewContainer;

/** Resolver for mapping username (path variable) to respective {@link UserDto} object */
@Component
@RequiredArgsConstructor
public class UserResolver implements PathVariableArgumentResolver {

  private final UserFacade userFacade;

  @Override
  public boolean supportsParameter(MethodParameter parameter) {
    return parameter.getParameterType().equals(UserDto.class);
  }

  @Override
  public Object resolveArgument(
      @NotNull MethodParameter parameter,
      ModelAndViewContainer mavContainer,
      @NotNull NativeWebRequest webRequest,
      WebDataBinderFactory binderFactory) {
    String username = getPathVariable(Mappings.PLACEHOLDER_USER, webRequest);

    return userFacade.resolveUser(username);
  }
}

File: src/main/kotlin/com/github/njuro/jard/rewrite/user/UserResolver.kt
package com.github.njuro.jard.rewrite.user

import com.github.njuro.jard.rewrite.common.PLACEHOLDER_USER
import com.github.njuro.jard.rewrite.user.dto.UserDto
import com.github.njuro.jard.utils.PathVariableArgumentResolver
import org.springframework.core.MethodParameter
import org.springframework.stereotype.Component
import org.springframework.web.bind.support.WebDataBinderFactory
import org.springframework.web.context.request.NativeWebRequest
import org.springframework.web.method.support.ModelAndViewContainer

/** Resolver for mapping username (path variable) to respective [UserDto] object  */
@Component
class UserResolver(private val userFacade: UserFacade) : PathVariableArgumentResolver {

    override fun supportsParameter(parameter: MethodParameter): Boolean =
        parameter.parameterType == UserDto::class.java


    override fun resolveArgument(
        parameter: MethodParameter,
        mavContainer: ModelAndViewContainer?,
        webRequest: NativeWebRequest,
        binderFactory: WebDataBinderFactory?
    ): UserDto {
        val username = getPathVariable(PLACEHOLDER_USER, webRequest)
        return userFacade.resolveUser(username)
    }
}
